File: AocCsharp/Day06/Day06.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using AocCsharp.Utils;

namespace AocCsharp.Day06;

public static class Day06
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day06_test.txt");
        var realInputPath = Path.Combine("..", "inputs", "day06.txt");

        TestRunner.RunSolution("Part 1", Part1, testInputPath, realInputPath, expectedTestResult: 4277556);
        TestRunner.RunSolution("Part 2", Part2, testInputPath, realInputPath, expectedTestResult: 3263827);
    }

    public static long Part1(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0) return 0;

        int height = lines.Length;

        // Find max width without LINQ
        int width = 0;
        for (int i = 0; i < lines.Length; i++)
        {
            if (lines[i].Length > width)
                width = lines[i].Length;
        }

        // Pad lines to ensure they are all the same length
        var grid = new string[height];
        for (int i = 0; i < height; i++)
        {
            grid[i] = lines[i].PadRight(width);
        }

        var problems = new List<(List<long> Numbers, char Op)>();
        
        int? startCol = null;

        for (int col = 0; col < width; col++)
        {
            bool isEmptyCol = true;
            for (int row = 0; row < height; row++)
            {
                if (grid[row][col] != ' ')
                {
                    isEmptyCol = false;
                    break;
                }
            }

            if (!isEmptyCol)
            {
                if (startCol == null)
                {
                    startCol = col;
                }
            }
            else
            {
                if (startCol != null)
                {
                    // End of a block
                    problems.Add(ParseProblem(grid, startCol.Value, col - 1));
                    startCol = null;
                }
            }
        }

        // Handle last block if it extends to the edge
        if (startCol != null)
        {
            problems.Add(ParseProblem(grid, startCol.Value, width - 1));
        }

        long total = 0;
        foreach (var p in problems)
        {
            long result = p.Numbers[0];
            for (int i = 1; i < p.Numbers.Count; i++)
            {
                if (p.Op == '+') result += p.Numbers[i];
                else if (p.Op == '*') result *= p.Numbers[i];
            }
            total += result;
        }

        return total;
    }

    private static (List<long> Numbers, char Op) ParseProblem(string[] grid, int startCol, int endCol)
    {
        var numbers = new List<long>();
        char op = ' ';

        int height = grid.Length;

        // Numbers are in all rows except the last
        for (int row = 0; row < height - 1; row++)
        {
            // Find first and last non-space characters in the range
            int start = -1;
            int end = -1;
            for (int col = startCol; col <= endCol; col++)
            {
                if (grid[row][col] != ' ')
                {
                    if (start == -1) start = col;
                    end = col;
                }
            }

            if (start != -1)
            {
                var span = grid[row].AsSpan(start, end - start + 1);
                if (long.TryParse(span, out long num))
                {
                    numbers.Add(num);
                }
            }
        }

        // Operator is in the last row
        for (int col = startCol; col <= endCol; col++)
        {
            char c = grid[height - 1][col];
            if (c == '+' || c == '*')
            {
                op = c;
                break;
            }
        }

        return (numbers, op);
    }

    public static long Part2(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0) return 0;

        int height = lines.Length;

        // Find max width without LINQ
        int width = 0;
        for (int i = 0; i < lines.Length; i++)
        {
            if (lines[i].Length > width)
                width = lines[i].Length;
        }

        // Pad lines to ensure they are all the same length
        var grid = new string[height];
        for (int i = 0; i < height; i++)
        {
            grid[i] = lines[i].PadRight(width);
        }

        var problems = new List<(List<long> Numbers, char Op)>();

        int? startCol = null;

        for (int col = 0; col < width; col++)
        {
            bool isEmptyCol = true;
            for (int row = 0; row < height; row++)
            {
                if (grid[row][col] != ' ')
                {
                    isEmptyCol = false;
                    break;
                }
            }

            if (!isEmptyCol)
            {
                if (startCol == null)
                {
                    startCol = col;
                }
            }
            else
            {
                if (startCol != null)
                {
                    // End of a block - parse reading right-to-left
                    problems.Add(ParseProblemRightToLeft(grid, startCol.Value, col - 1));
                    startCol = null;
                }
            }
        }

        // Handle last block if it extends to the edge
        if (startCol != null)
        {
            problems.Add(ParseProblemRightToLeft(grid, startCol.Value, width - 1));
        }

        long total = 0;
        foreach (var p in problems)
        {
            long result = p.Numbers[0];
            for (int i = 1; i < p.Numbers.Count; i++)
            {
                if (p.Op == '+') result += p.Numbers[i];
                else if (p.Op == '*') result *= p.Numbers[i];
            }
            total += result;
        }

        return total;
    }

    private static (List<long> Numbers, char Op) ParseProblemRightToLeft(string[] grid, int startCol, int endCol)
    {
        var numbers = new List<long>();
        char op = ' ';

        int height = grid.Length;

        // Read each column from right to left
        for (int col = endCol; col >= startCol; col--)
        {
            // Build number directly from digits without allocations
            long num = 0;
            bool hasDigits = false;

            // Read digits from top to bottom in this column (rows 0 to height-2, excluding operator row)
            for (int row = 0; row < height - 1; row++)
            {
                char c = grid[row][col];
                if (c != ' ')
                {
                    num = num * 10 + (c - '0');
                    hasDigits = true;
                }
            }

            if (hasDigits)
            {
                numbers.Add(num);
            }
        }

        // Operator is in the last row - find it anywhere in this block
        for (int col = startCol; col <= endCol; col++)
        {
            char c = grid[height - 1][col];
            if (c == '+' || c == '*')
            {
                op = c;
                break;
            }
        }

        return (numbers, op);
    }
}

File: AocC/day06.c
#include "day06.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>

void day06_run(void) {
    const char* test_path = "../inputs/day06_test.txt";
    const char* real_path = "../inputs/day06.txt";

    run_solution("Part 1", day06_part1, test_path, real_path, 4277556);
    run_solution("Part 2", day06_part2, test_path, real_path, 3263827);
}

typedef struct {
    int64_t numbers[100];
    int number_count;
    char op;
} Problem;

static Problem parse_problem(char** grid, int height, int start_col, int end_col) {
    Problem p;
    p.number_count = 0;
    p.op = ' ';

    // Numbers are in all rows except the last
    for (int row = 0; row < height - 1; row++) {
        // Find first and last non-space characters in the range
        int start = -1, end = -1;
        for (int col = start_col; col <= end_col; col++) {
            char c = grid[row][col];
            if (c != ' ') {
                if (start == -1) start = col;
                end = col;
            }
        }

        if (start != -1) {
            // Extract number
            char buf[32];
            int len = 0;
            for (int col = start; col <= end && len < 31; col++) {
                buf[len++] = grid[row][col];
            }
            buf[len] = '\0';

            int64_t num;
            if (sscanf(buf, "%lld", (long long*)&num) == 1) {
                p.numbers[p.number_count++] = num;
            }
        }
    }

    // Operator is in the last row
    for (int col = start_col; col <= end_col; col++) {
        char c = grid[height - 1][col];
        if (c == '+' || c == '*') {
            p.op = c;
            break;
        }
    }

    return p;
}

static Problem parse_problem_rtl(char** grid, int height, int start_col, int end_col) {
    Problem p;
    p.number_count = 0;
    p.op = ' ';

    // Read each column from right to left
    for (int col = end_col; col >= start_col; col--) {
        // Build number from digits in this column (top to bottom, excluding operator row)
        int64_t num = 0;
        bool has_digits = false;

        for (int row = 0; row < height - 1; row++) {
            char c = grid[row][col];
            if (c != ' ' && isdigit((unsigned char)c)) {
                num = num * 10 + (c - '0');
                has_digits = true;
            }
        }

        if (has_digits) {
            p.numbers[p.number_count++] = num;
        }
    }

    // Operator is in the last row
    for (int col = start_col; col <= end_col; col++) {
        char c = grid[height - 1][col];
        if (c == '+' || c == '*') {
            p.op = c;
            break;
        }
    }

    return p;
}

int64_t day06_part1(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    // Parse lines - use dynamic allocation for large inputs
    char** lines = (char**)malloc(1000 * sizeof(char*));
    int height = 0;

    char* line = strtok(data, "\r\n");
    while (line && height < 1000) {
        lines[height++] = line;
        line = strtok(NULL, "\r\n");
    }

    if (height == 0) {
        free(lines);
        free(data);
        return 0;
    }

    // Find max width
    int width = 0;
    for (int i = 0; i < height; i++) {
        int len = (int)strlen(lines[i]);
        if (len > width) width = len;
    }

    // Pad lines to same width - dynamic allocation
    char** grid = (char**)malloc(height * sizeof(char*));
    for (int i = 0; i < height; i++) {
        grid[i] = (char*)malloc(width + 1);
        int len = (int)strlen(lines[i]);
        memcpy(grid[i], lines[i], len);
        for (int j = len; j < width; j++) {
            grid[i][j] = ' ';
        }
        grid[i][width] = '\0';
    }

    // Find problem blocks - dynamic allocation
    Problem* problems = (Problem*)malloc(10000 * sizeof(Problem));
    int problem_count = 0;
    int start_col = -1;

    for (int col = 0; col < width; col++) {
        bool is_empty = true;
        for (int row = 0; row < height; row++) {
            if (grid[row][col] != ' ') {
                is_empty = false;
                break;
            }
        }

        if (!is_empty) {
            if (start_col == -1) start_col = col;
        } else {
            if (start_col != -1) {
                problems[problem_count++] = parse_problem(grid, height, start_col, col - 1);
                start_col = -1;
            }
        }
    }

    if (start_col != -1) {
        problems[problem_count++] = parse_problem(grid, height, start_col, width - 1);
    }

    // Calculate total
    int64_t total = 0;
    for (int i = 0; i < problem_count; i++) {
        Problem* p = &problems[i];
        if (p->number_count == 0) continue;

        int64_t result = p->numbers[0];
        for (int j = 1; j < p->number_count; j++) {
            if (p->op == '+') result += p->numbers[j];
            else if (p->op == '*') result *= p->numbers[j];
        }
        total += result;
    }

    // Cleanup
    for (int i = 0; i < height; i++) {
        free(grid[i]);
    }
    free(grid);
    free(problems);
    free(lines);
    free(data);

    return total;
}

int64_t day06_part2(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    // Parse lines - use dynamic allocation for large inputs
    char** lines = (char**)malloc(1000 * sizeof(char*));
    int height = 0;

    char* line = strtok(data, "\r\n");
    while (line && height < 1000) {
        lines[height++] = line;
        line = strtok(NULL, "\r\n");
    }

    if (height == 0) {
        free(lines);
        free(data);
        return 0;
    }

    // Find max width
    int width = 0;
    for (int i = 0; i < height; i++) {
        int len = (int)strlen(lines[i]);
        if (len > width) width = len;
    }

    // Pad lines to same width - dynamic allocation
    char** grid = (char**)malloc(height * sizeof(char*));
    for (int i = 0; i < height; i++) {
        grid[i] = (char*)malloc(width + 1);
        int len = (int)strlen(lines[i]);
        memcpy(grid[i], lines[i], len);
        for (int j = len; j < width; j++) {
            grid[i][j] = ' ';
        }
        grid[i][width] = '\0';
    }

    // Find problem blocks - dynamic allocation
    Problem* problems = (Problem*)malloc(10000 * sizeof(Problem));
    int problem_count = 0;
    int start_col = -1;

    for (int col = 0; col < width; col++) {
        bool is_empty = true;
        for (int row = 0; row < height; row++) {
            if (grid[row][col] != ' ') {
                is_empty = false;
                break;
            }
        }

        if (!is_empty) {
            if (start_col == -1) start_col = col;
        } else {
            if (start_col != -1) {
                problems[problem_count++] = parse_problem_rtl(grid, height, start_col, col - 1);
                start_col = -1;
            }
        }
    }

    if (start_col != -1) {
        problems[problem_count++] = parse_problem_rtl(grid, height, start_col, width - 1);
    }

    // Calculate total
    int64_t total = 0;
    for (int i = 0; i < problem_count; i++) {
        Problem* p = &problems[i];
        if (p->number_count == 0) continue;

        int64_t result = p->numbers[0];
        for (int j = 1; j < p->number_count; j++) {
            if (p->op == '+') result += p->numbers[j];
            else if (p->op == '*') result *= p->numbers[j];
        }
        total += result;
    }

    // Cleanup
    for (int i = 0; i < height; i++) {
        free(grid[i]);
    }
    free(grid);
    free(problems);
    free(lines);
    free(data);

    return total;
}
